Optimized dev:urn-catalog:generate for PHPStorm

JeroenBoersma opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

When generating the `urn-catalog` from within any docker setup with path mappings this isn't reflected correctly to PHPStorm because the path's won't match up. This is no problem for people running local php instances.

### Description (*)
This change resolves the project directory and replaces all paths to `$PROJECT_DIR$`,  this also makes moving your project possible. The change does a compare from the realpath for the given `.idea/misc.xml` to resolve the projectdir.

### Fixed Issues (if relevant)
I didn't find any, so it only bothered me.

### Manual testing scenarios (*)
Run in any console `bin/magento dev:urn-catalog:generate test.xml` this will replace the absolute paths to `$PROJECT_DIR/path/to/file.xsd` in the location attribute.

### Questions or comments
No Unittests because there weren't any to optimize.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

JeroenBoersma committed:
Removed unused parameter from constructor

JeroenBoersma committed:
Resolve $PROJECT_DIR$ when generating PHPStorm xml

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=23291) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=23291) before we can accept your contribution.<br/>

JeroenBoersma committed:
Processed phpcs feedback

- removed unused import
- removed unused @param

JeroenBoersma commented:
```
PHP Code Sniffer detected 1 violation(s): 

FILE: .../app/code/Magento/Developer/Model/XmlCatalog/Format/PhpStorm.php
----------------------------------------------------------------------
FOUND 0 ERRORS AND 2 WARNINGS AFFECTING 1 LINE
----------------------------------------------------------------------
 137 | WARNING | The use of function realpath() is discouraged
 137 | WARNING | The use of function dirname() is discouraged
----------------------------------------------------------------------

```

This is intended behavior. Please help me howto proceed?

sidolov commented on app/code/Magento/Developer/Model/XmlCatalog/Format/PhpStorm.php:
```diff
@@ -129,4 +125,27 @@ private function initEmptyFile(\DOMDocument $dom)
+     * @param string $configFilePath
+     * @return string
+     */
+    public function resolveProjectPath($configFilePath): string
```
Make method private

sidolov commented on app/code/Magento/Developer/Model/XmlCatalog/Format/PhpStorm.php:
```diff
@@ -129,4 +125,27 @@ private function initEmptyFile(\DOMDocument $dom)
+     * @param string $xsdPath
+     * @return string
+     */
+    public function getFileLocationInProject(string $ideaDir, string $xsdPath): string
```
Please, make method private

sidolov commented on app/code/Magento/Developer/Model/XmlCatalog/Format/PhpStorm.php:
```diff
@@ -10,18 +10,15 @@
-     */
-    private $currentDirRead;
+    const PROJECT_PATH_IDENTIFIER = '$PROJECT_DIR$';
+    const IDEA_PATH = '.idea';
```
Please, make constants private

JeroenBoersma replied:
ok, fair enough

sidolov commented on app/code/Magento/Developer/Model/XmlCatalog/Format/PhpStorm.php:
```diff
@@ -34,16 +31,13 @@ class PhpStorm implements FormatInterface
      * @param DomDocumentFactory $domDocumentFactory
      */
     public function __construct(
-        ReadFactory $readFactory,
```
Unfortunately, we are not able to remove dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

JeroenBoersma replied:
check, missed out on that  one, did a crawl on code, didn't find any ;)

JeroenBoersma committed:
Revert "Processed phpcs feedback"

This reverts commit 211fe95022e5d49cda66278a26c74632344349f3.

JeroenBoersma committed:
Revert "Removed unused parameter from constructor"

This reverts commit 0db89b2dd4645827dc05536af0bbfc17db7b0a43.

JeroenBoersma committed:
Converted internals to private methods

* internally used code is no longer public

JeroenBoersma committed:
Second iteration use re-added readDirectory relative path

Use currentReaddir to make relative paths.
Second iteration which uses way less code with the same result.

- removed realpath usage
- removed dirname usage
- removed unused method
- removed unused constant

JeroenBoersma commented:
Did a second iteration with fresh mind, came up with a "better" solution.
Thanks for the feedback

magento-engcom-team commented:
Hi @sivaschenko, thank you for the review.
[ENGCOM-5404](https://jira.corp.magento.com/browse/ENGCOM-5404) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

magento-engcom-team commented:
@JeroenBoersma thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

engcom-Alfa commented:
:heavy_check_mark: QA Passed
